File: POA2.java
import java.util.Arrays;
import java.util.List;

class Usuario {
    String nombre;
    boolean autenticado;
    String rol;

    public Usuario(String nombre, boolean autenticado, String rol) {
        this.nombre = nombre;
        this.autenticado = autenticado;
        this.rol = rol;
    }
}

interface AccionConfidencial {
    void ejecutar(Usuario usuario);
}

class RequiereAutorizacion implements AccionConfidencial {
    private List<String> rolesPermitidos;
    private AccionConfidencial accion;

    public RequiereAutorizacion(List<String> rolesPermitidos, AccionConfidencial accion) {
        this.rolesPermitidos = rolesPermitidos;
        this.accion = accion;
    }

    @Override
    public void ejecutar(Usuario usuario) {
        if (!usuario.autenticado) {
            System.out.println("Usuario " + usuario.nombre + " no está autenticado.");
            return;
        }

        if (!rolesPermitidos.contains(usuario.rol)) {
            System.out.println("Usuario " + usuario.nombre + " no tiene permiso para acceder a esta función.");
            return;
        }

        accion.ejecutar(usuario); 
    }
}

class VerDatosConfidenciales implements AccionConfidencial {
    @Override
    public void ejecutar(Usuario usuario) {
        System.out.println("Acceso concedido a " + usuario.nombre + " para ver datos confidenciales.");
    }
}

public class POA2 {
    public static void main(String[] args) {
        Usuario usuarioAdmin = new Usuario("Carlos", true, "admin");
        Usuario usuarioEmpleado = new Usuario("Luis", true, "empleado");
        Usuario usuarioNoAutenticado = new Usuario("Ana", false, "supervisor");

        AccionConfidencial verDatos = new RequiereAutorizacion(
            Arrays.asList("admin", "supervisor"),
            new VerDatosConfidenciales()
        );

        verDatos.ejecutar(usuarioAdmin);
        verDatos.ejecutar(usuarioEmpleado);
        verDatos.ejecutar(usuarioNoAutenticado);
    }
}

File: POA2.py
def requiere_autorizacion(roles_permitidos):
    def decorador(func):
        def wrapper(usuario, *args, **kwargs):
            if not usuario.get('autenticado'):
                print(f"Usuario {usuario['nombre']} no está autenticado.")
                return
            
            if usuario.get('rol') not in roles_permitidos:
                print(f"Usuario {usuario['nombre']} no tiene permiso para acceder a esta función.")
                return
            
            return func(usuario, *args, **kwargs)
        
        return wrapper
    return decorador

@requiere_autorizacion(roles_permitidos=['admin', 'supervisor'])
def ver_datos_confidenciales(usuario):
    print(f"Acceso concedido a {usuario['nombre']} para ver datos confidenciales.")

usuario_admin = {'nombre': 'Carlos', 'autenticado': True, 'rol': 'admin'}
usuario_empleado = {'nombre': 'Luis', 'autenticado': True, 'rol': 'empleado'}
usuario_no_autenticado = {'nombre': 'Ana', 'autenticado': False, 'rol': 'supervisor'}

# Ejecuciones
ver_datos_confidenciales(usuario_admin)            
ver_datos_confidenciales(usuario_empleado)         
ver_datos_confidenciales(usuario_no_autenticado)   